astropy/nddata/mixins/ndarithmetic.py

```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
+
+        kwargs:
+            Any other parameter that should be passed to the
+            different :meth:`NDArithmeticMixin._arithmetic_data` (or wcs, ...)
+            methods.
+            # TODO: Add Example for a use-case with subclassing.
 
         Returns
         -------
-        result : `~astropy.nddata.NDData`
-            The resulting dataset
-
-        Notes
-        -----
-        This method requires the datasets to have identical WCS
-        properties, equivalent units, and identical shapes.
-        Meta-data get set to None in the resulting dataset. The unit
-        in the result is the same as the unit in ``self``. Uncertainties
-        are propagated, although correlated errors are not supported
-        by any of the built-in uncertainty classes.  If uncertainties
-        are assumed to be correlated, a warning is issued by default
-        (though this can be disabled via the
-        ``astropy.nddata.conf.warn_unsupported_correlated``
-        configuration item). Values masked in either dataset before
-        the operation are masked in the resulting dataset.
+        result: `~numpy.ndarray` or `~astropy.units.Quantity`
+            The resulting data as array (in case both operands were without
+            unit) or as quantity if at least one had a unit.
+
+        kwargs: `dict`
+            the kwargs should contain all the other attributes (besides data
+            and unit) to create a new instance for the result. Creating the
+            new instance is up to the calling method, for example
+            :meth:`NDArithmeticMixin.add`.
+
+            TODO: Reword this? (embray)
+            I did formulate a new text, does this look better to you?
+
         """
+        # Convert the operand to the same class this allows for arithmetic
+        # operations with numbers, lists, numpy arrays, numpy masked arrays
+        # astropy quantities, masked quantities and of other subclasses of
+        # NDData
+        operand = self.__class__(operand)
+
+        # TODO: Check that both data have numeric dtypes otherwise we could
+        # end up with some arithmetic operation on string types. (init only
+        # enforces that the dtype is NOT objects)?
+
+        kwargs = {}
+        # First check that the WCS allows the arithmetic operation
+        if compare_wcs is None:
+            kwargs['wcs'] = None
+        elif not compare_wcs:
+            if self.wcs is None:
+                kwargs['wcs'] = deepcopy(operand.wcs)
+            else:
+                kwargs['wcs'] = deepcopy(self.wcs)
+        else:
+            kwargs['wcs'] = self._arithmetic_wcs(operand, compare_wcs,
+                                                 **kwds)
+        # Then calculate the resulting data (which can but not needs to be a
+        # quantity)
+        result = self._arithmetic_data(operation, operand, **kwds)
+        # Determine the other properties
+        if propagate_uncertainties is None:
```

Comment:
@wkerzendorf - My `Variable` class (#3715) is quite complete and does track uncertainties properly for all numpy `ufunc`. I've not worked on it further since I posted it, however, since there really should be a bit of general thought in when to stop tracking correlations.